Add autocapitalize attribute

rlanday opened the pull request:
The Chrome team is currently attempting to update our implementation of the autocapitalize attribute in Chrome for Android (currently a nonstandard extension introduced by Apple) to match the behavior of iOS Safari (specifically, to add support for autocapitalize on editable regions, inheritance from the form owner for <input> and <textarea> elements, and some of the IDL behavior). As part of this work, we would like to standardize this attribute in the HTML spec.

The goal with this spec change is to document iOS Safari's behavior (so ideally, Apple won't have to make any changes to their implementation) so other browsers (such as Chrome for Android) can implement the attribute with the same behavior.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/3273/acknowledgements.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">/acknowledgements.html</a>  ( <a href="https://whatpr.org/html/3273/dff66be...6257fd7/acknowledgements.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">diff</a> )
<a href="https://whatpr.org/html/3273/dom.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">/dom.html</a>  ( <a href="https://whatpr.org/html/3273/dff66be...6257fd7/dom.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">diff</a> )
<a href="https://whatpr.org/html/3273/index.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">/index.html</a>  ( <a href="https://whatpr.org/html/3273/dff66be...6257fd7/index.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">diff</a> )
<a href="https://whatpr.org/html/3273/indices.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">/indices.html</a>  ( <a href="https://whatpr.org/html/3273/dff66be...6257fd7/indices.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">diff</a> )
<a href="https://whatpr.org/html/3273/interaction.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">/interaction.html</a>  ( <a href="https://whatpr.org/html/3273/dff66be...6257fd7/interaction.html" title="Last updated on Dec 15, 2017, 11:11 PM GMT (6257fd7)">diff</a> )

annevk commented:
Thank you for taking this on, it's much appreciated! This will help resolve https://www.w3.org/Bugs/Public/show_bug.cgi?id=28228, a longstanding bug.

How does this relate to the `inputmode` work in #3260? Presumably if you specify both there needs to be some conflict resolution.

cc @dtapuska

annevk commented:
Even though this has shipped already, it might be worth it to get some insights from @whatwg/a11y and @whatwg/i18n for possible future enhancements.

xfq commented:
Some opinions from the i18n WG:

* As Addison [said](https://github.com/w3c/html/issues/208#issuecomment-346975167), among other things, there are [annoying corner cases](https://w3c.github.io/charmod-norm/#definitionCaseFolding) in case folding.

* As Richard [said](https://github.com/w3c/html/issues/208#issuecomment-347270249), different languages may need different default values, although Addison's (and Terence's) opinion in that issue was the attribute may not need to be language or input method–dependent, because it's only a hint.

For more information, please see also https://github.com/w3c/html/issues/208 and https://github.com/w3c/html/issues/1092. Thank you.

dtapuska commented:
@annevk I don't believe there really is any conflict resolution for auto capitalize and what keyboard is shown... inputmode doesn't specify if the first letter should be capital. The only real conflict that comes up is if the keyboard is shown to be a numeric keyboard but that is exactly the same case as showing a default keyboard and typing the first entry as a number.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+  <code data-x="">characters</code> maps to the <i>characters</i> state. The keywords <code
+  data-x="">on</code> and <code data-x="">sentences</code> map to the <i>sentences</i> state,
+  which is also the <i data-x="invalid value default">invalid value default</i>. The <i
+  data-x="missing value default">missing value default</i> is a separate <i>default</i> state.</p>
```
I would prefer we make a table from the above two paragraphs and also included a brief description. See https://html.spec.whatwg.org/#attr-button-type for precedent. That way the meaning of the values are also somewhat clear in the developer's edition.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+  data-x="missing value default">missing value default</i> is a separate <i>default</i> state.</p>
+
+  <dl class="domintro">
+
```
Please remove this newline.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+   <dt><var>element</var> . <code subdfn data-x="dom-autocapitalize">autocapitalize</code> [ = <var>value</var> ]</dt>
+
+   <dd>
+
```
As well as this newline.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+
+    <p>Can be set, to set the <code data-x="attr-autocapitalize">autocapitalize</code> content
+    attribute (and thereby change the autocapitalization behavior for the element).</p>
+
```
And this one.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+    attribute (and thereby change the autocapitalization behavior for the element).</p>
+
+   </dd>
+
```
And this one.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+  attribute</dfn>" algorithm) on the element:<p>
+
+  <ol>
+
```
Also this newline.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+
+  <ol>
+
+    <li><p>If the <code data-x="attr-autocapitalize">autocapitalize</code> content attribute is
```
This is indented one space too many. We use a single space.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+    owner</span>.</p></li>
+
+    <li><p>Return an empty string.</p></li>
+
```
This newline can go.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+  methods are allowed to conditionally ignore or not support) have the following meanings:</p>
+
+  <ul>
+
```
This newline can go.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+    other letters should default to lowercase.</p></li>
+
+    <li><p><i>characters</i>: All letters should default to uppercase.</p></li>
+
```
This newline can go.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+  <p>The <i>used autocapitalization hint</i> for an element is computed using the following algorithm:</p>
+
+  <ol>
+
```
This newline can go.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+
+  <ol>
+
+    <li><p>If the element is an <code>input</code> element whose <code
```
One space too many again.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
+    standard), the <i>used autocapitalization hint</i> is not defined for this element, and this
+    specification does not define what autocapitalization behavior should be used when entering text
+    into this element.</p></li>
+
```
This newline can go.

othermaciej commented:
Tagging @cdumez, @rniwa 

I don't think Apple is opposed to standardizing this. The reason it's on iOS Safari but not macOS is that  iOS native interface features autocapitalization in text controls by default, but macOS doesn't. It makes sense to provide author override on iOS, but on macOS it would have no effect. 

iOS autocaplitization takes the form of making the keyboard's sticky shift key be default-on for letters where the user is expected to want a capital. This defaults to on (in sentence mode) in most text-entry contexts. The user can tap shift to choose not to capitalize the letter after all. There's no analogue for this on platforms that have a physical keyboard such as macOS. macOS does have an autocorrect

I think we don't even expose the IDL interface aspects of this on macOS or other desktop-oriented ports of WebKit, but that is likely more for historical reasons than because of being against it on the merits. We could expose the IDL attribute, have it reflect the markup attribute, and still make it have no material effect.

I have not verified whether the PR's requirements match iOS Safari behavior as intended. I am not sure which of the details and quirks described here are required for compatibility with mobile-targeted web content, either.

r12a commented:
[1]

I may have missed it, but reading the PR source i didn't see a clear description of what autocapitalization does.  This seems to have lead to some confusion in some quarters. I see 

> assist users by automatically capitalizing the first letter of sentences

but i suspect this is slightly incorrect, since my understanding is that the browser doesn't capitalise the first letter of a sentence, but instead, as  @othermaciej says above, it sends a hint to the browser or system about how to display the visual keyboard - ie. it doesn't alter what the user typed at all. Ie. for `autocapitalize=sentence` the keyboard switches to show uppercase letters, but the user is quite at liberty to switch the keyboard back to lowercase before typing anything, so the end result is not necessarily an uppercase letter.

Is my understanding correct?  If i do have the right end of the stick here, it may be worth making that clearer (unless i missed it).

(This is handy actually, since you don't need to worry about language-specific rules for upper and lowercasing, such as whether the i needs to be a Turkish capital letter I with dot or not, which is a nest of snakes (if not familiar with that see http://rishida.net/blog/201512.html#20160907). If this is a prompt to the keyboard display, the turkish user is likely to be using a turkish keyboard and will be presented with the appropriate character.)

[2]

[OT] (i hope) The autocomplete behaviour appears to mimick the keys typed by the user in order to apply case. I assume that that is a separate thing, and it takes its cues for what to case convert from what the user typed, and that it is aware of the need to follow local case conversion rules (though i don't know whether it gets language cues for that from the system language, or from the language of the page content).

[3] 

It would be useful to be able to set an autocapitalize default at quite a high level in the document, so that it is inherited by controls that can use it, but can be overridden where necessary. One of the reasons that might be useful is...

[4] ... that some pages may want to set a default that is different from the norm, without having to set a special value on every control on the page.

For example,  Cherokee may require `autocapitalize="characters"` to be the default (for the time being), because of the font used, given that lowercase Cherokee letters have only recently been added to Unicode (that may change over time, but a good number of fonts have not yet added them).

The Georgian default should presumably be `autocapitalize="none"` by default for pages written in Georgian, given that modern Georgian text uses only lowercase letters (even though uppercase characters are encoded in the Georgian block). 

(For a little more information about Cherokee and Georgian, see http://rishida.net/blog/201512.html#20161106).


hth

rlanday commented:
@xfq I'm not familiar with the particular case folding issues, but my assumption is that input methods will apply the autocapitalization hint in a way that makes sense for each language. As @othermaciej noted for iOS, the way autocapitalization is currently implemented for a character (on both iOS and Android) is by making the shift key default to "on" for the next character to be typed, so I don't think we run into issues with changing the character to something the user isn't expecting.

I think if different languages should default to different autocapitalization behavior, that's something that the input method can do in response to the hint (e.g. having a language-specific behavior for the "default" state).


@annevk, @othermaciej I did notice that macOS Safari doesn't expose the IDL interface. I'm not sure how critical exposing the interface is for platforms that don't support the autocapitalize feature, but I suppose it would be good for cross-platform consistency? Chrome currently does expose the IDL the same way on all platforms, but there are cases where the IDL values are different from what Safari would provide (which I intend to fix once we agree on the desired behavior).


@r12a 

[1] I could add a note in the spec as to exactly how autocapitalization is currently implemented by on-screen keyboards (by toggling the sticky shift key). I think there are probably some other cases though like swipe input and voice input where it also makes sense to respect the attribute.

[2] I don't think there's really significant interaction with autocomplete here. I think that e.g. if sentence capitalization is enabled, and someone tries to start a sentence with "iPhone" but it comes out "IPhone" due to the autocapitalization, it's fine for autocomplete to correct that to "iPhone," despite the first letter no longer being capitalized. Autocomplete behavior is typically based on the language the current input method is configured with (for example, on Android, you can switch between different languages of Gboard)

[3], [4] This is an interesting idea; I'm not sure if demand has come up yet in practice though. The particular cases with Cherokee and Georgian seem better handled by the input method choosing to use only uppercase or lowercase letters as the default behavior and/or completely ignoring the autocapitalize attribute if appropriate.

annevk commented:
I think it would make sense to expose the `autocapitalize` IDL attribute always to make things easier for developers (then in the future they won't have to distinguish between undefined and a string value), but perhaps we change the processing model of the getter slightly so it always returns the empty string on platforms where this hint is not applicable? Or perhaps make that even more general, and say to return the empty string if the given value is not supported.

(Aside: anyone else having a hard time typing this attribute?)

domenic commented:
One thing to keep in mind is that it's not a clear mobile/desktop split, as you can use software keyboards even on desktop platforms (e.g. I do so on my Windows 10 tablets when they are undocked). Or you can dock hardware keyboards onto phones and tablets. Indeed, you can use both at once.

Also note that the processing model outlined already has a mismatch between the "used autocapitalization hint" and the value of `el.autocapitalize`, for various reasons. So it wouldn't be too bad to have `el.autocapitalize` have some value that is meaningless on situations with hardware keyboards.

That said I don't have a real preference for what to do in such situations and am happy to defer to implementers.

annevk commented on source:
```diff
@@ -74438,6 +74440,148 @@ body { display:none }
 
+  <h4>Autocapitalization</h4>
+
+  <p>Some methods of entering text, for example voice input and on-screen keyboards on mobile
```
Please use "virtual keyboard" throughout.

rlanday commented:
I've updated the pull request to respond to the comments about formatting (I also added cross-referencing links to the attribute index), and I've tried to add some more clarity around how the attribute is typically implemented.

I haven't added anything about what the behavior should be if the attribute isn't supported. I suspect that desktop Safari might eventually want to add support e.g. to allow enabling/disabling autocapitalization for voice input (Chrome for Android currently support's the attribute for Gboard's voice input), so I don't think it makes a lot of sense to talk about "platforms where the hint is not applicable"; rather, it depends on the input method.

I will start working on the web platform tests. @othermaciej, please let us know if the Safari team would like to request changes to the behavior in the spec (or if I haven't properly documented Safari's current behavior; this would hopefully be caught once I start writing tests though).

domenic commented:
I got a chance to work on this and update the model a bit to flow better, e.g. with less conversion between states and attributes, and a single unified "autocapitalization hint" concept. It was meant to be no normative changes, except one fix mentioned below. This brought up a couple questions:

- You stated "empty value default", but there is no such thing; there is instead "invalid value default". I made that change, which means that `autocapitalize="asdf"` will end up in the sentences state. Is that correct?
- Is this step correct, in computing the element's "own autocapitalization hint"? "If the autocapitalize content attribute is present on element, **and its value is not the empty string,** return the state of the attribute." In practice, this means the mapping of `<input autocapitalize>` just ignores the attribute (since its value is empty).
- You state that the last algorithm step of "used autocapitalization hint" cannot be reached. If so, we can clean that up to something shorter. But by my reading of the editing spec, that's not the case. In particular, there are plenty of elements that are neither input/textarea elements, nor editing hosts, nor editable elements. For example, `<span contenteditable=false>` should end up in step 4. Similarly, any disconnected node, such as `document.createElement("span")`, will reach step 4. Maybe step 4 should instead say something like "If this step is reached, the element is not editable, so it has no used autocapitalization hint." Or did I misread the editing spec?

Thanks for the great PR!

rlanday commented:
Thanks for the edits!

- Oops! Yes, I think you‘ve fixed it properly. The missing value default is the default state and the invalid value default is the sentences state.

autocapitalize="asdf" should indeed end up in the sentences state.

- iOS Safari currently treats empty values (e.g. <input autocapitalize>) as if the attribute is not present at al, so that’s what I currently have in the spec; however, the Apple documentation claims empty values are supposed to be treated the same as “sentences” (at least as far as enabling capitalization; it’s not really clear about the IDL value):
https://developer.apple.com/library/content/documentation/AppleApplications/Reference/SafariHTMLRef/Articles/Attributes.html#//apple_ref/doc/uid/TP40008058-autocapitalize

I found a case where the Apple documentation is clearly wrong (in fact, I think form inheritance is the _only_ case where an empty attribute value might ever be expected to actually change the behavior):

<form autocapitalize="none">
    <input autocapitalize>
</form>

In this case, in iOS Safari, the <input> element inherits the “none” state.

I currently have a WebKit bug open about this spec issue since the Apple documentation does not match Safari’s actual behavior:
https://bugs.webkit.org/show_bug.cgi?id=179133

Chrome for Android does not currently support inheriting autocapitalize from the form owner, and the default is sentences anyway, so we haven’t really made a decision about how to interpret an empty value. We agree with Safari that invalid values should trigger “sentences” (in fact, this is how we currently support “on”; we don’t explicitly look for the keyword).

- I wrote that the algorithm should be run “during text input into an element” (is this the correct terminology?). You are correct that step 4 can clearly be reached if we can run the algorithm on arbitrary elements. If we only run the algorithm on elements undergoing text input, then I don’t think it can be reached.

domenic commented:
> I believe both Chrome and Safari treat an empty string the same way as omitting the attribute entirely.

I think I was missing this detail, as I just naively assumed `<input autocapitalize>` was one way of saying "turn autocapitalization on for this input". If that hasn't ever worked in Chrome or Safari, then the current spec seems correct. Thanks for catching me!

I also missed the "during text input" bit. With that in mind, we can just simplify it I think; no need to hedge things about the future. E.g.:

> Assert: this step is never reached, since text input only occurs in elements that meet one of the above criteria.

rlanday commented:
Ok, I simplified the last step.

I think we're still waiting on feedback from Safari implementers? I will include the web platform tests in the Chromium CL, which I will upload once we're in agreement about the rest of the spec addition.

annevk commented on source:
```diff
@@ -74438,6 +74442,173 @@ body { display:none }
+  using the following algorithm:</p>
+
+  <ol>
+    <li><p>If <var>element</var> is an <code>input</code> element whose <code
```
Indentation is off here. Should be three spaces for all list items. (@domenic or I can probably fix this nit when everything else is sorted.)

annevk commented:
> Also note that the processing model outlined already has a mismatch between the "used autocapitalization hint" and the value of `el.autocapitalize`, for various reasons.

I was wondering what these reasons were. Compatibility concerns perhaps? It seems preferable to expose the used value in the IDL attribute, if we can, though this would require modifying that computation somewhat since you'd be able to reach step 4. (At least, in my impression IDL attributes generally tend to reflect what ends up being used, if it's not too expensive.)

domenic commented:
Yeah, that would be too far from what's implemented today, is my understanding.

rlanday commented:
There are two "mismatches" between the "used autocapitalization hint" and the value of `el.autocapitalize`. The first is that the hint doesn't match the IDL value for password, URL, and email <input>s.

Safari (whose behavior I wrote up in the spec) treats <input> elements of all types the same as far as IDL value. I think it passes this value to the keyboard, together with the input type, and the keyboard decides never to enable autocapitalization for password, URL, or email inputs.

There are three other possible behaviors I can think of:

- <input> elements of password, URL, or email always have an IDL value of "" (if the attribute is value empty or the attribute is not present) or "none" (if any other value is specified). This complicates the IDL algorithm by adding a dependency on the <input> type, and it's kind of weird to return "" for value empty/not present, but "none" for an overridden value. It seems the IDL attribute should either reflect what's specified, or what the used hint will be, and not some weird combination.

- <input> elements of password, URL, or email always have an IDL value of "none", regardless of whether or not the autocapitalize attribute is specified. This introduces an inconsistency in that <input> elements of other types return "" when the attribute is not specified. We could change the behavior for all <input> types so that the IDL attribute is always the used value. I think this is basically Chrome's current behavior. Note that this behavior seems to make it impossible to specify to an IME that we want the "default" behavior (the default has to be chosen by the browser-side code), so I think it's suboptimal.

- We support autocapitalize on password, URL, and email <input> elements the same way we do for search and text inputs. Chrome currently does this (although it's disabled by default). Chrome currently does not support inheriting the value of <autocapitalize> from <form> elements. For compatibility reasons with Safari, password, URL, and email inputs shouldn't ever inherit a non-none value of autocapitalize from their form owners. So we'd have to complicate how inheritance works to do this. The Chrome implementers believe that autocapitalization is fairly useless for these input types and would like to remove our support to match Safari, rather than complicate form inheritance.


The second "mismatch" is that for text input into an editable region, we use the IDL value from the editing host, and not any descendant element we may be inserting the text into. From one perspective, this isn't really a "mismatch," but rather just being specific about which element we're using the property from. I think it's fairly standard to get properties like autocapitalize from the editing host (e.g. I think autocomplete, autocorrect, and spellcheck are implemented the same way). It's not really feasible to try to set up a system of inheritance where any element in an editable region can override the value of autocapitalize for itself and its descendants, since a selection can span multiple elements, and then we wouldn't know which value we should use.

annevk commented:
Thanks, I think keeping what we have now is okay. Remainder:

 - [ ] web-platform-tests PR
 - [ ] Bugs filed against Edge, Firefox, and Safari (macOS primarily, as I understand it)

@cdumez or @rniwa confirming @othermaciej's statements would be great, but I'm reasonably confident this'll work out.

(I'm assuming Chrome is already tracking this.)

rlanday commented:
Chrome bug: https://bugs.chromium.org/p/chromium/issues/detail?id=776618
Edge bug: https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/15102197/
Firefox bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1425291
WebKit bug: https://bugs.webkit.org/show_bug.cgi?id=180821

I will work on getting the test cases finished.

rlanday commented:
I uploaded a test as part of my Chromium CL:
https://chromium-review.googlesource.com/c/chromium/src/+/830415/1/third_party/WebKit/LayoutTests/external/wpt/html/editing/editing-0/autocapitalization/autocapitalize.html

annevk commented:
@domenic want to do the honors?

domenic commented on source:
```diff
@@ -10058,6 +10058,7 @@ interface <dfn>HTMLElement</dfn> : <span>Element</span> {
   readonly attribute DOMString <span data-x="dom-accessKeyLabel">accessKeyLabel</span>;
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-draggable">draggable</span>;
   [<span>CEReactions</span>] attribute boolean <span data-x="dom-spellcheck">spellcheck</span>;
+  [<span>CEReactions</span>] attribute boolean <span data-x="dom-autocapitalize">autocapitalize</span>;
```
This should be DOMString, not boolean. I'll fix.

sideshowbarker commented:
Also now supported in https://checker.html5.org/ and https://validator.w3.org/nu/

sideshowbarker commented:
Updated MDN to add `autocapitalize` as a global attribute:

https://developer.mozilla.org/en-US/docs/Web/HTML/Global_attributes/autocapitalize
